core/src/main/java/org/apache/calcite/sql/SqlOperator.java

```diff
@@ -510,7 +510,7 @@ public RelDataType deriveType(
     final SqlOperator sqlOperator =
         SqlUtil.lookupRoutine(validator.getOperatorTable(), getNameAsId(),
             argTypes, null, null, getSyntax(), getKind(),
-            validator.getCatalogReader().nameMatcher());
+            validator.getCatalogReader().nameMatcher(), false);
```

Comment:
probably define a version with default parameter set as false

Reply:
A better way should be the flag of whether validator allows the type coercion.